modules/bmsBidAdapter.js

```diff
@@ -0,0 +1,153 @@
+      floor.currency === DEFAULT_CURRENCY
+    ) {
+      return floor.floor;
+    }
+  }
+  return null;
+}
+
+const converter = ortbConverter({
+  context: {
+    netRevenue: true, // Default net revenue configuration
+    ttl: 100, // Default time-to-live for bid responses
+  },
+  imp,
+  request,
+});
+
+function request(buildRequest, imps, bidderRequest, context) {
+  let request = buildRequest(imps, bidderRequest, context);
+
+  // Add publisher ID
+  deepSetValue(request, 'site.publisher.id', context.publisherId);
+  return request;
+}
+
+function imp(buildImp, bidRequest, context) {
+  let imp = buildImp(bidRequest, context);
+  const floor = getBidFloor(bidRequest);
+  imp.tagid = bidRequest.params.placementId;
+
+  if (floor) {
+    imp.bidfloor = floor;
+    imp.bidfloorcur = DEFAULT_CURRENCY;
+  }
+
+  return imp;
+}
+
+export const spec = {
+  code: BIDDER_CODE,
+  gvlid: GVLID,
+  supportedMediaTypes: [BANNER],
+
+  // Validate bid request
+  isBidRequestValid: function (bid) {
+    return !!bid.params.placementId && !!bid.params.publisherId;
+  },
+
+  // Build OpenRTB requests using `ortbConverter`
+  buildRequests: function (validBidRequests, bidderRequest) {
+    const context = {
+      publisherId: validBidRequests.find(
+        (bidRequest) => bidRequest.params?.publisherId
+      )?.params.publisherId,
+    };
+
+    const ortbRequest = converter.toORTB({
+      bidRequests: validBidRequests,
+      bidderRequest,
+      context,
+    });
+
+    // Add extensions to the request
+    ortbRequest.ext = ortbRequest.ext || {};
+    deepSetValue(ortbRequest, 'ext.gvlid', GVLID);
+
+    return [
+      {
+        method: 'POST',
+        url: ENDPOINT_URL,
+        data: ortbRequest,
+        options: {
+          contentType: 'application/json',
+        },
+      },
+    ];
+  },
+
+  interpretResponse: (serverResponse) => {
+    if (!serverResponse || isEmpty(serverResponse.body)) return [];
+
+    let bids = [];
+    serverResponse.body.seatbid.forEach((response) => {
+      response.bid.forEach((bid) => {
+        const mediaType = bid.ext?.mediaType || 'banner';
+        bids.push({
+          ad: replaceAuctionPrice(bid.adm, bid.price),
+          adapterCode: BIDDER_CODE,
+          cpm: bid.price,
+          creativeId: bid.ext.bms.adId,
+          creative_id: bid.ext.bms.adId,
+          currency: serverResponse.body.cur || 'USD',
+          deferBilling: false,
+          deferRendering: false,
+          width: bid.w,
+          height: bid.h,
+          mediaType,
+          netRevenue: true,
+          originalCpm: bid.price,
+          originalCurrency: serverResponse.body.cur || 'USD',
+          requestId: bid.impid,
+          seatBidId: bid.id,
+          ttl: 1200,
+        });
+      });
+    });
+    return bids;
+  },
+
+  onBidWon: function (bid) {
+    const { burl, nurl } = bid || {};
+
+    if (nurl) {
+      triggerPixel(replaceAuctionPrice(nurl, bid.originalCpm || bid.cpm));
```

Comment:
I recommend you do this macro substitution server side, also sendbeacon or fetch keep alive may outlive the page view